Source/LogModifier.swift

```diff
@@ -24,10 +24,140 @@
+    /// - Returns: Modified message string.
+    func modifyMessage(_ message: String, with logLevel: LogLevel, attributes: [String: Any]) -> String
+    
+    /// - Parameter message: The message to be modified.
+    /// - Parameter context: The context captured when this log message was recorded.
+    /// - Parameter attributes: The array of message attributes that were part of the original log message.
+    /// - Returns: Modified message string.
+    func modifyMessage(_ message: String, with context: LogMessageContext, attributes: [String: Any]) -> String
+}
+
+public extension LogModifier {
+    /// Default implementation is provided for existing clients of Willow. It ignores the attributes.
+    func modifyMessage(_ message: String, with logLevel: LogLevel, attributes: [String: Any]) -> String {
+        return modifyMessage(message, with: logLevel)
+    }
+    
+    /// Default implementation is in terms of original modifyMessage method.
+    func modifyMessage(_ message: String, with context: LogMessageContext, attributes: [String: Any]) -> String {
+        return modifyMessage(message, with: context.logLevel)
+    }
+}
+
+// MARK: -
+
+/// The `PropertyExpansionModifier` replaces property references in the message string with values found in the `attributes` or `context`.
+///
+/// For example, if the message was `"Failure reason: {attributes.reason}"` and the passed attributes array contained
+/// a key of `reason` with the value `Service unavailable`, the modified message would be `"Failure reason: Service unavailable"`
+///
+/// The syntax for an attribute property reference is: `{attributes.(key)}`.  You replace the `(key)` with an actual key value from the attributes array.
+/// The syntax for a context property reference is: `{context.(key)}`. In this case, the `(key)` must be one of the following values:
+///
+/// * `logLevel` - The log level
+/// * `timestamp` - The timestamp in milliseconds since epoch.
+/// * `file` - The name of the file where the log message was recorded.
+/// * `file.lastPathComponent` - The optional name of the file where the log message was recorded, but only the final path component.
+/// * `function`- The name of the function in which the log message was recorded.
+/// * `line` - The line number where the log message was recorded.
+/// * `subsystem` - The optional application-defined subsystem this log message belongs to.
+/// * `category` - The optional application-defined category this log message belongs to.
+/// * `session` - The optional application-defined session identifier.
+/// * `user` - The optional application-defined user identifier.
+///
+/// Some of these context values are optional. If they are not provided in the log message context, the property reference
+/// will be replaced with the string `missing`.
+///
+/// References are case sensitive. The replacement will fail if the cases do not match.
+///
+open class PropertyExpansionModifier: LogModifier {
```

Comment:
This is a handy modifier that makes it easy to include properties from the attributes or context in a message directly in the message itself.